sdk/core/azure-core/src/main/java/com/azure/core/util/Contexts.java

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.implementation.http.HttpPipelineCallContextHelper;
+
+import java.util.Objects;
+
+/**
+ * A utility type that can be used to add and retrieve instances commonly used in {@link Context}.
+ */
+public final class Contexts {
+
+    private static final String PROGRESS_REPORTER_CONTEXT_KEY = "com.azure.core.util.ProgressReporter";
+
+    private Context context;
+
+    private Contexts(Context context) {
+        this.context = Objects.requireNonNull(context, "'context' must not be null");
+    }
+
+    /**
+     * Creates {@link Contexts} from empty {@link Context}.
+     * @return The {@link Contexts} instance.
+     */
+    public static Contexts empty() {
+        return with(Context.NONE);
+    }
+
+    /**
+     * Creates {@link Contexts} from supplied {@link Context}.
+     * @param context Existing {@link Context}. Must not be null.
+     * @return The {@link Contexts} instance.
+     * @throws NullPointerException If {@code context} is null.
+     */
+    public static Contexts with(Context context) {
+        return new Contexts(context);
+    }
+
+    /**
+     * Creates {@link Contexts} from supplied {@link HttpPipelineCallContext}.
+     * @param context Existing {@link HttpPipelineCallContext}. Must not be null.
+     * @return The {@link Contexts} instance.
+     * @throws NullPointerException If {@code context} is null.
+     */
+    public static Contexts with(HttpPipelineCallContext context) {
+        return new Contexts(HttpPipelineCallContextHelper.getContext(context));
+    }
+
+    /**
+     * Adds {@link ProgressReporter} instance to the {@link Context}.
+     * @param progressReporter The {@link ProgressReporter} instance.
+     * @return Itself.
+     */
+    public Contexts setProgressReporter(ProgressReporter progressReporter) {
```

Comment:
How does the user use this when working with an async client? 

This will need creating Azure context first, then converting to reactor context and then internally will be converted to Azure context again.

Should we make the context key (PROGRESS_REPORTER_CONTEXT_KEY) public, so that users can directly set the key, value in reactor context? 

```java
Context azureContext = Contexts.empty().setProgressReporter(reporter).getContext();
reactor.util.context.Context reactorContext = FluxUtil.toReactorContext(azureContext);
asyncClient.foo().contextWrite(reactorContext);
```

Reply:
I did this in the prototype
https://github.com/kasobol-msft/azure-sdk-for-java/blob/7c65a6c06ec9845ebf394077ff83e92be0950dfe/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java#L1010-L1017
Which is waste of allocation as you say.
We should make well known keys public constants. Or perhaps we should also create `ReactorContexts` that mirrors `Contexts`.

If we're to define public constants where this would be? We don't have a good place for this, so things like `azure-eagerly-read-response` are scattered across packages.

Reply:
We could define all these public constants in `Contexts` class itself. The other option is to define them in `Context` like we do for `Configuration`.